File: cpp/wordbreak.cpp
#include <memory>
#include <iostream>
#include <string>
#include <unordered_map>
#include <vector>

using namespace std;
class node {
    public:
        node():complete{false} {}
        unordered_map<char, node* > ms;
        bool complete;
        ~node() {}
};

class Trie {
public:

    node* root;

    Trie(): root{new node()} {}
    void add(string& s)
    {
        int i = 0;
        int len = s.size();
        //
        node* n = root;
        while(i < len)
        {
            auto it = n->ms.find(s[i]);
            if (it == n->ms.end())
                break;
            n = it->second;
            ++i;
        }

        while(i < len)
        {
            node* temp{new node()};
            n->ms[s[i]] = temp;
            n = temp;
            ++i;
        }

        n->complete = true;
    }
};


class Solution {
public:

    vector<string> wordBreak(string s, vector<string>& wordDict) {
        Trie trie;

        for (auto& e : wordDict)
        {
            trie.add(e);
        }

        unordered_map<int, vector<string>> mm;
        vector<string> r;

        foo(trie, s, 0, r, mm);

        return r;
    }


    void foo(Trie& trie,
            string& s,
            int idx,
            vector<string>& r,
            unordered_map<int, vector<string>>& mm)
     {
        if (idx == s.size())
        {
            r.push_back("");
            return;
        }

        // use cache
        if (mm.find(idx) != mm.end())
        {
            r = mm[idx];
            return;
        }

        node* n = trie.root;
        int i = 0;

        string tt;
        vector<string> rr;
        while (n->ms.find(s[idx + i]) != n->ms.end())
        {
            n = n->ms[s[idx + i]];
            tt += s[idx + i];

            if (n->complete)
            {
                foo(trie, s, idx + i + 1, rr, mm);
                for (auto& e : rr)
                {
                    if (e.size() > 0)
                        r.push_back(tt + " " + e);
                    else
                        r.push_back(tt);
                }
                rr.clear();
            }
            ++i;
        }
        mm[idx] = r;
    }
};
int main()
{
    Solution s;

    vector<string> v {"cat","cats","and","sand","dog"};
    auto r = s.wordBreak("catsanddog", v);

    for (auto&e : r)
    {
        cout << e << endl;
    }
}

File: java/wordBreak.java
import java.util.HashMap;
import java.util.Map;
import java.lang.Character;
import java.lang.Integer;
import java.util.List;
import java.util.ArrayList;
import java.util.Arrays;


class Node {
    Node() {}
    public Map<Character, Node> ms = new HashMap<Character, Node>();
    public boolean complete = false;
}

class Trie {
    public Node root = new Node();
    public void add(String s)
    {
        int len = s.length();
        Node n = root;
        int i = 0;
        while (i < len)
        {
            Character key = Character.valueOf(s.charAt(i));
            if (n.ms.containsKey(key))
            {
                n = n.ms.get(key);
                ++i;
            }
            else
            {
                break;
            }
        }

        while (i < len)
        {
            Node val = new Node();
            Character key = Character.valueOf(s.charAt(i));

            n.ms.put(key, val);
            // System.out.println("add : " + key.toString());
            n = val;
            ++i;
        }
        n.complete = true;
        // System.out.println("done for : " + s);
    }
}

class Solution {
    public List<String> wordBreak(String s, List<String> wordDict) {
        List<String> l = new ArrayList<String>();
        Trie trie = new Trie();
        for (String e : wordDict)
        {
            trie.add(e);
        }
        Map<Integer, List<String>> mm = new HashMap<>();

        foo(trie, s, 0, l, mm);

        for (Map.Entry e : mm.entrySet()) //
            System.out.println(e.getKey() + " : " + e.getValue());

        return l;
    }

    void foo(Trie trie, String s, int idx, List<String> r,
            Map<Integer, List<String>> mm)
    {
        if (idx == s.length())
        {
            r.add(new String());
            return;
        }

        if (mm.containsKey(Integer.valueOf(idx)))
        {
            r.addAll(mm.get(idx));
            return;
        }

        Node n = trie.root;
        int i = 0;
        Character key = Character.valueOf(s.charAt(idx + i));
        StringBuffer tt = new StringBuffer();
        while (n.ms.containsKey(key))
        {
            n = n.ms.get(key);
            tt.append(key);
            // System.out.println("> " + tt.toString());

            if (n.complete)
            {
                List<String> rr = new ArrayList<>();

                foo(trie, s, idx + i + 1, rr, mm);
                for (String e: rr)
                {
                    if (e.length() == 0)
                    {
                        r.add(tt.toString());
                    }
                    else
                    {
                        r.add(tt.toString() + " " + e);
                    }
                }
            }
            ++i;
            if (idx + i == s.length())
                break;
            key = Character.valueOf(s.charAt(idx + i));
            // System.out.println(">> " + key.toString());
        }

        mm.put(Integer.valueOf(idx), r);
        System.out.println("> " + String.valueOf(idx));
        for (String e : r)
            System.out.println(e);

    }

    public static void main(String args[])
    {
        Solution sol = new Solution();

        String[] sa = {"cat", "cats", "and", "sand", "dog"};
        List<String> wordDict = new ArrayList<>(Arrays.asList(sa));
        List<String> l = sol.wordBreak("catsanddog", wordDict);

        for (String e: l)
        {
            System.out.println(e);
        }

        System.out.println("done");

    }
}
